python/lsst/ts/mtdome/mock_llc/mock_motion/shutter_motion.py

```diff
@@ -0,0 +1,343 @@
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = [
+    "ShutterMotion",
+    "CLOSED_POSITION",
+    "NUM_MOTORS_PER_SHUTTER",
+    "OPEN_POSITION",
+    "SHUTTER_SPEED",
+]
+
+import logging
+import math
+from typing import Tuple
+
+from ...enums import LlcMotionState
+from .base_llc_motion import BaseLlcMotion
+
+# The number of motors per shutter.
+NUM_MOTORS_PER_SHUTTER = 2
+
+# The shutter is 0% open.
+CLOSED_POSITION = 0.0
+# The shutter is 100% open.
+OPEN_POSITION = 100.0
+# The shutter speed (%/s). This is an assumed value such that the shutter opens
+# or closes in 10 seconds.
+SHUTTER_SPEED = 10.0
+
+
+class ShutterMotion(BaseLlcMotion):
+    """Simulator for one door (of two) of the aperture shutter motion of the
+    MTDome.
+
+    Parameters
+    ----------
+    start_position: `float`
+        The initial position [%].
+    start_tai: `float`
+        The current TAI time, unix seconds. To model the real dome, this should
+        be the current time. However, for unit tests it can be convenient to
+        use other values.
+    """
+
+    def __init__(
+        self,
+        start_position: float,
+        start_tai: float,
+    ):
+        super().__init__(
+            start_position=start_position,
+            min_position=CLOSED_POSITION,
+            max_position=OPEN_POSITION,
+            max_speed=SHUTTER_SPEED,
+            start_tai=start_tai,
+        )
+        self.log = logging.getLogger("ShutterMotion")
+        # Keep track of being in error state or not.
+        self.motion_state_in_error = False
+        # Keep track of which drives are in error state.
+        self.drives_in_error_state = [False] * NUM_MOTORS_PER_SHUTTER
+        # Keep track of when motion state went into ERROR
+        self._error_start_tai = 0.0
+        # Keep track of the position at the moment that ERROR state starts.
+        self._error_state_position = 0.0
+
+    def set_target_position_and_velocity(
+        self,
+        start_tai: float,
+        end_position: float,
+        crawl_velocity: float,
+        motion_state: LlcMotionState,
+    ) -> float:
+        raise NotImplementedError(
```

Comment:
Would it be practical to to use this method instead of creating a new set_shutter_target_position_and_velocity, and raise an exception if crawl_velocity != 0?

It seems a pity to have a different custom method for setting position and velocity for every subsystem that does not support crawling.

Using this existing method may make more sense if motion_state can have a useful default value, since then you can also provide a default value of 0 for crawl_velocity. (It would be unfortunate to make users explicitly specify crawl_velocity=0 when commanding a system that cannot crawl).

Reply:
This is a great idea and I implemented that.